packages/core/src/external/ehr/job/create-resource-diff-bundles/steps/contribute/ehr-contribute-resource-diff-bundles-direct.ts

```diff
@@ -0,0 +1,285 @@
+      if (reportError) {
+        await setJobEntryStatus({
+          ...entryStatusParams,
+          entryStatus: "failed",
+        });
+      }
+      throw error;
+    }
+    if (this.waitTimeInMillis > 0) await sleep(this.waitTimeInMillis);
+  }
+}
+
+async function getEhrOnlyResourcesFromS3({
+  ehr,
+  cxId,
+  metriportPatientId,
+  ehrPatientId,
+  resourceType,
+  jobId,
+}: Omit<FetchBundleParams, "bundleType">): Promise<Resource[]> {
+  const bundle = await fetchBundle({
+    ehr,
+    cxId,
+    metriportPatientId,
+    ehrPatientId,
+    resourceType,
+    bundleType: BundleType.RESOURCE_DIFF_EHR_ONLY,
+    jobId,
+  });
+  if (!bundle?.bundle.entry || bundle.bundle.entry.length < 1) return [];
+  return bundle.bundle.entry.flatMap(entry => {
+    if (!entry.resource) return [];
+    return [entry.resource];
+  });
+}
+
+async function hydrateEhrOnlyResources({
+  ehr,
+  tokenId,
+  cxId,
+  practiceId,
+  metriportPatientId,
+  ehrPatientId,
+  ehrOnlyResources,
+}: {
+  ehr: EhrSource;
+  tokenId: string | undefined;
+  cxId: string;
+  practiceId: string;
+  metriportPatientId: string;
+  ehrPatientId: string;
+  ehrOnlyResources: Resource[];
+}): Promise<Resource[]> {
+  const hydratedEhrOnlyResources = [...ehrOnlyResources];
+  const fetchedResourceIds = new Set([
+    ...hydratedEhrOnlyResources.flatMap(resource => resource.id ?? []),
+  ]);
+  for (let i = 0; i < hydrateEhrOnlyResourceAttempts; i++) {
+    const { missingReferences } = getReferencesFromResources({
+      resourcesToCheckRefs: hydratedEhrOnlyResources,
+    });
+    if (missingReferences.length < 1) break;
+    const hydrationArgs = uniqBy(missingReferences, "id");
+    const hydrationErrors: { error: unknown; id: string; type: string }[] = [];
+    await executeAsynchronously(
+      hydrationArgs,
+      async (params: ReferenceWithIdAndType) => {
+        const { id, type } = params;
+        if (fetchedResourceIds.has(id)) return;
+        try {
+          const bundle = await getResourceBundleByResourceId({
+            ehr,
+            ...(tokenId && { tokenId }),
+            cxId,
+            practiceId,
+            metriportPatientId,
+            ehrPatientId,
+            resourceType: type,
+            resourceId: id,
+            useCachedBundle: true,
+          });
+          const resource = bundle.entry?.[0]?.resource;
+          if (!resource) return;
+          hydratedEhrOnlyResources.push(resource);
+        } catch (error) {
+          if (error instanceof BadRequestError || error instanceof NotFoundError) return;
+          const refToString = JSON.stringify(params);
+          log(`Failed to hydrate resource ${refToString}. Cause: ${errorToString(error)}`);
+          hydrationErrors.push({ error, ...params });
+        } finally {
+          fetchedResourceIds.add(id);
+        }
+      },
+      {
+        numberOfParallelExecutions: parallelRequests,
+        delay: delayBetweenRequestBatches.asMilliseconds(),
```

Comment:
Wouldn't be better to use min/max jitter? It adds some delay and spreads the load over time